photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala

```diff
@@ -104,28 +104,39 @@ object RandomEffectCoordinate {
       randomEffectOptimizationProblem: RandomEffectOptimizationProblem[Function],
       randomEffectModel: RandomEffectModel): (RandomEffectModel, Option[RandomEffectOptimizationTracker]) = {
 
-    val updatedModels = randomEffectDataSet
+    val updatedModelsAndTrackers = randomEffectDataSet
       .activeData
       .join(randomEffectOptimizationProblem.optimizationProblems)
       .join(randomEffectModel.modelsRDD)
       .mapValues {
         case (((localDataSet, optimizationProblem), localModel)) =>
           val trainingLabeledPoints = localDataSet.dataPoints.map(_._2)
+          val updatedModel = optimizationProblem.run(trainingLabeledPoints, localModel)
+          val stateTrackers = optimizationProblem.getStatesTracker
 
-          optimizationProblem.run(trainingLabeledPoints, localModel)
+          (updatedModel, stateTrackers)
       }
+      .setName(s"Updated models and state trackers for random effect ${randomEffectDataSet.randomEffectType}")
+      .persist(StorageLevel.VERY_FREQUENT_REUSE_RDD_STORAGE_LEVEL)
     val updatedRandomEffectModel = randomEffectModel
-      .update(updatedModels)
-      .setName(s"Updated random effect model")
-    val optimizationTracker = if (randomEffectOptimizationProblem.isTrackingState) {
-        val stateTrackers = randomEffectOptimizationProblem.optimizationProblems.map(_._2.getStatesTracker.get)
+      .update(updatedModelsAndTrackers.mapValues(_._1))
+      .setName(s"Updated models for random effect ${randomEffectDataSet.randomEffectType}")
+      .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
+      .materialize()
+    val optimizationTracker: Option[RandomEffectOptimizationTracker] =
+      if (randomEffectOptimizationProblem.isTrackingState) {
+        val stateTrackers = updatedModelsAndTrackers.map(_._2._2.get)
         val randomEffectTracker = new RandomEffectOptimizationTracker(stateTrackers)
-          .setName(s"Random effect optimization tracker")
+          .setName(s"State trackers for random effect ${randomEffectDataSet.randomEffectType}")
           .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
           .materialize()
 
         Some(randomEffectTracker)
-      } else None
+      } else {
+        None.asInstanceOf[Option[RandomEffectOptimizationTracker]]
+      }
+
+    updatedModelsAndTrackers.unpersist()
```

Comment:
It looks like the 3 loops are tied together by `updatedModelsAndTrackers`. Since this is unpersisted, can we use the model and the tracker(s) right where we need them in the first loop, instead of having 3 loops, and persisting in between? Presumably that would be faster, and possibly simpler?